Node maintenance window enhancement

vchlum opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Bug/feature Description
* The new feature provides a special reservation that can be easily used for maintenance. The reservations can be submitted only on entire host(s) and they are "prioritized" over common reservations. This feature is only for managers and operators.

#### Affected Platform(s)
* All

#### Cause / Analysis / Design
* **[design document](https://pbspro.atlassian.net/wiki/spaces/PD/pages/879493121/Node+maintenance+window+enhancement)**
* **[pbspro community forum](http://community.pbspro.org/t/allowing-to-schedule-node-maintenance-with-a-possibility-to-run-new-jobs-until-the-maintenance-begins/1306)**


#### Solution Description
* The special reservation is submitted with extend parameter, which is acl restricted. Once the reservation is submitted, the command pbs_rsub immediately confirms the reservation. The confirmation process recognizes reservation prefix and detects all overlapping reservation. The overlapping reservations are degraded (other reservations are altered in order not to overlap the special reservation) and the scheduler will try to reconfirm them later. Already running reservations are not reconfirmed - just degraded.

#### Testing logs/output
* [ptl_reservations_output.txt](https://github.com/PBSPro/pbspro/files/3029400/ptl_reservations_output.txt)
* [ptl_maintenance_output.txt](https://github.com/PBSPro/pbspro/files/3029401/ptl_maintenance_output.txt)



#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [x] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

bhroam commented on test/tests/functional/pbs_maintenance_reservations.py:
```diff
@@ -0,0 +1,192 @@
+        Test if the maintenance reservation is denied for common user
+        """
+        a = {'--hosts': self.mom.shortname}
+        r = Reservation(TEST_USER, attrs=a)
```
You are making an assumption that the key to attrs can be an rsub option.  The interface says they're attributes.  Since rsub itself is doing a lot of work for this feature, I'd use run_cmd(). (and other places in the test plan)

vchlum replied:
It makes sense, done.

bhroam commented on test/tests/functional/pbs_maintenance_reservations.py:
```diff
@@ -0,0 +1,192 @@
+        self.server.submit(r2)
+
+        exp_attr = {'reserve_state': (MATCH_RE, 'RESV_DEGRADED|10'),
+                    'reserve_substate': 12}
```
I thought maintenance reservations marked overlapping reservations in conflict instead of degraded.  Also check for the shortened resv_nodes

vchlum replied:
The state is DEGRADED. Only the substate is IN_CONFLICT, which is the '12'. It is possible to set also the state to IN_CONFLICT (just for being more obvious), but it is not necessary for the feature. Should I set also the state to IN_CONFLICT?

bhroam replied:
Oh, I thought degraded was a substate.  The reservation was in state confirmed or running, but the substate was degraded.  This would degrade the reservation.  I guess I'm not as familiar with the degraded reservation code as I thought.

bhroam commented on test/tests/functional/pbs_maintenance_reservations.py:
```diff
@@ -0,0 +1,192 @@
+        self.server.expect(RESV, exp_attr, id=rid1)
+
+        a2 = {'reserve_start': now + 3602,
+              'reserve_end': now + 3604,
```
Could we use longer reservations than 5 seconds?  Since they're an hour out it shouldn't matter.  We've run into weird race conditions with short timeframes.  They shouldn't happen here, but it can't hurt to make these longer.

vchlum replied:
OK, I use much longer intervals now.

bhroam commented on test/tests/functional/pbs_maintenance_reservations.py:
```diff
@@ -0,0 +1,192 @@
+
+        exp_attr = {'reserve_state': (MATCH_RE, 'RESV_CONFIRMED|2'),
+                    'resv_nodes': '(%s:ncpus=1)' % self.momB.shortname}
+        self.server.expect(RESV, exp_attr, id=rid1)
```
Consider more tests (either some or all of these ideas).  Your maintenance reservation starts after the overlapping reservation and end before it.  Write one where the maintenance reservation starts before the overlapping reservation and ends during it.  Another where the maintenance reservation starts during the overlapping reservation and ends after it.  The third is where the maintenance reservation starts before and ends after the overlapping reservation.   Write a test where the two reservations overlap in nodes but not time and make sure the normal reservation is not marked degraded.  Write an end to end test.  Submit a maintenance reservation, see it overlap with another reservation, and make sure they work properly when they are running.  Write one where you see an overlapping reservation get reconfirmed after it is degraded.

vchlum replied:
Thank you for suggestions. Several new tests have been added. I also added tests on top of suggested.

vchlum commented on src/cmds/pbs_rsub.c:
```diff
@@ -696,8 +765,154 @@ main(int argc, char *argv[], char *envp[])
+			exit(pbs_errno);
+		}
+
+		extend = strdup("m");
```
It is freed on line 958.

vchlum commented on src/server/req_rescq.c:
```diff
@@ -222,6 +222,165 @@ cnvrt_timer_init()
+				/* remove possible trailing '+' */
+				tmp_len = strlen(presv->ri_wattr[(int)RESV_ATR_resv_nodes].at_val.at_str);
+				if (presv->ri_wattr[(int)RESV_ATR_resv_nodes].at_val.at_str[tmp_len - 1] == '+')
+					presv->ri_wattr[(int)RESV_ATR_resv_nodes].at_val.at_str[tmp_len - 1] = '\0';
```
This is after the modification of resv_nodes and it covers the cases: "+(node2:whatever)+(node3:whatever)" and "(node1:whatever)+(node2:whatever)+", but I realized that I forgot "(node1:whatever)++(node3:whatever)".

I also realized that "select=1:ncpus=2" and "select=2" produce different resv_nodes on the same multi-noded host. "select=1:ncpus=2" produces: "resv_nodes = (vn[0]:ncpus=1+vn[1]:ncpus=1)" but "select=2" produces: "resv_nodes = (vn[0]:ncpus=1)+(vn[1]:ncpus=1)". There was a bug in this.

I fixed the issues in the commit "bug fix in changing resv_nodes". A commit actually addressing your comments will come soon.

bhroam replied:
I'm confused how resv_nodes can contain a ++.  The case I was talking about was a chunk between two others (vn1:ncpus=1)+(vn2:ncpus=1)+(vn3:ncpus=1).  The middle chunk has a plus on both sides.  If you remove them both, the chunks will slide together.

vchlum replied:
The modification of resv_nodes works like this: First, the string "(vn[n]:whatever)" is removed from resv_nodes but no '+' is removed yet - only the chunk itself. The extra '+' is removed afterwards.

So in your example, it means:
- For removing vn1 we get: "+(vn2:ncpus=1)+(vn3:ncpus=1)"
- For removing vn2 we get: "(vn1:ncpus=1)++(vn3:ncpus=1)"
- For removing vn3 we get: "(vn1:ncpus=1)+(vn2:ncpus=1)+"

After this, the extra '+' is removed:
- First, the resv_nodes leading '+' is checked and removed (not the chunk leading, but the whole resv_nodes leading - this is engaged when vn1 is removed). 
- Second, the resv_nodes trailing '+' is checked and removed (not the chunk trailing, but the whole resv_nodes trailing - this is engaged when vn3 is removed). 
- Third, the resv_nodes is searched for double '+' and if found, one of the plus is removed (this is engaged when vn2 is removed).

I will add some simple comment to the code.

bhroam replied:
Oh OK, I understand now.  Yes a code comment would be appreciated.

vchlum commented on src/server/svr_jobfunc.c:
```diff
@@ -4131,6 +4131,10 @@ change_enableORstart(resc_resv *presv, int which, char *value)
+	if (which == Q_CHNG_START && strcmp(value, ATR_TRUE) == 0 &&
+		! presv->ri_wattr[RESV_ATR_resv_nodes].at_flags & ATR_VFLAG_SET)
+		return (0);
+
```
It means: Never start a reservation queue if its reservation is without RESV_ATR_resv_nodes.

If such a reservation queue is started (the associated RESV_ATR_resv_nodes is not set) then the associated jobs could run on any node!

It covers the case when the maintenance wipes the whole RESV_ATR_resv_nodes of common reservation and the reservation is degraded but it will start anyway.

bhroam replied:
Ahh got it.  Thanks for the explanation.

vchlum commented on src/cmds/pbs_rsub.c:
```diff
@@ -696,8 +765,154 @@ main(int argc, char *argv[], char *envp[])
+
+				if (strcmp(bstat->name, *hostp) == 0 && atoi(ncpus) > 0) {
+					if (!select_str) {
+						select_str = (char*) malloc(strlen("select=host=") + strlen(bstat->name) + strlen(":ncpus=") +strlen(ncpus) + 1 + 1);
```
OK, fixed on all occurrences.

vchlum commented on src/cmds/pbs_rsub.c:
```diff
@@ -710,9 +925,43 @@ main(int argc, char *argv[], char *envp[])
+		char *resv_start_time_str;
+		time_t resv_start_time = 0;
+
+		reservid = strtok(new_resvname, " ");
```
Done.

vchlum commented on src/scheduler/resv_info.c:
```diff
@@ -736,7 +736,8 @@ query_resv(struct batch_status *resv, server_info *sinfo)
 	 */
 	if (advresv->resv->resv_state ==RESV_DEGRADED) {
```
Done

vchlum commented on src/server/req_rescq.c:
```diff
@@ -222,6 +222,165 @@ cnvrt_timer_init()
+	struct	resvinfo *rinfp, *prev;
+	attribute   tmpatr;
+
+	tmp_buf = (char *)malloc(sizeof(char) * strlen(pnode->nd_name) + 2);
```
OK, done.

vchlum commented on src/server/req_rescq.c:
```diff
@@ -222,6 +222,165 @@ cnvrt_timer_init()
+
+			if (presv->ri_giveback) {
+				/* resources were actually assigned to this reservation
+				 * we must accounted the resources back into the loaner's pool.
```
Done

vchlum commented on src/server/req_rescq.c:
```diff
@@ -222,6 +222,165 @@ cnvrt_timer_init()
+			end++;
+
+			if (presv->ri_giveback) {
+				/* resources were actually assigned to this reservation
```
OK, it is shorter now.

vchlum commented on src/server/req_rescq.c:
```diff
@@ -222,6 +222,165 @@ cnvrt_timer_init()
+	}
+
+	/* traverse the reservations of the node and remove the reservation if found */
+	for (prev=NULL, rinfp = pnode->nd_resvp; rinfp; prev=rinfp, rinfp = rinfp->next) {
```
Done

vchlum commented on test/tests/functional/pbs_maintenance_reservations.py:
```diff
@@ -0,0 +1,192 @@
+# trademark licensing policies.
+
+from tests.functional import *
+import time
```
OK, thanks, removed.

vchlum commented on test/tests/functional/pbs_maintenance_reservations.py:
```diff
@@ -0,0 +1,192 @@
+        except PbsSubmitError as err:
+            msg = err.msg[0].strip()
+
+        self.assertTrue("pbs_rsub: Unauthorized Request" == msg)
```
Changed across all tests.

vchlum commented:
Latest test logs:
[ptl_reservations_output2.txt](https://github.com/PBSPro/pbspro/files/3043022/ptl_reservations_output2.txt)
[ptl_maintenance_output2.txt](https://github.com/PBSPro/pbspro/files/3043023/ptl_maintenance_output2.txt)

bhroam commented on test/tests/functional/pbs_maintenance_reservations.py:
```diff
@@ -172,21 +395,120 @@ def test_maintenance_reconfirm_resv(self):
+        self.server.manager(MGR_CMD_SET, SERVER,
+                            {'managers': '%s@*' % TEST_USER})
+        self.server.manager(MGR_CMD_SET, SERVER,
+                            {'scheduler_iteration': 3})
```
Curious, why lower the scheduler_iteration?  The server should kick cycles when needed.

vchlum replied:
The same occurrence above was unnecessary (removed). But it is needed in the last two tests where the reconfirmation is done. If the reservation is degraded by maintenance reservation then no scheduler cycle is kicked. It will wait for the regular cycle. AFAIS, this is the same behavior as if the vnode of confirmed reservation turns down. Reconfirmation will also wait for the regular cycle.

vchlum replied:
Actually, only one occurrence is necessary. The rest is removed.

bhroam commented on test/tests/functional/pbs_maintenance_reservations.py:
```diff
@@ -172,21 +395,120 @@ def test_maintenance_reconfirm_resv(self):
+        jid1 = self.server.submit(j)
+
+        self.logger.info("Wait for the job to try to run (30 seconds)")
+        time.sleep(30)
```
Why are we waiting 30s for the job to run?  If the reservation is running, a cycle should be run when the job is submitted.

vchlum replied:
The job is expected to stay in Q and do not run because its reservation was fully overlapped. There could be a false positive result without any timeout. Maybe 10 seconds is enough, but I think 30 is bearable.

vchlum commented on test/tests/functional/pbs_maintenance_reservations.py:
```diff
@@ -36,60 +36,81 @@
+        end = self.now_plus_to_rsub_format(hours=2)
 
-        msg = ""
+        ret = self.submit_maintenance(TEST_USER, start, end,
```
I am sorry, I am not sure what do you mean by "have one reason the reservation can't be submitted". Please, could you clarify? Do you mean (for example) to always test whether the return code is not zero?

bhroam replied:
The test can also be rejected for permissions reasons.  Could you add TEST_USER to the managers line like you do in other tests?  This means the only reason the reservation can be rejected is due to -lselect and --hosts.

vchlum replied:
Thank you for clarification, done.

vchlum commented on test/tests/functional/pbs_maintenance_reservations.py:
```diff
@@ -135,27 +220,165 @@ def test_maintenance_degrade_reservation(self):
+    def test_maintenance_reconfirm_resvation_and_run(self):
+        """
+        Test if the overlapping reservation is reconfirmed correctly.
+        Wait for both reservation to start and submit jobs into this
```
Yes, fixed.

vchlum commented on test/tests/functional/pbs_maintenance_reservations.py:
```diff
@@ -135,27 +220,165 @@ def test_maintenance_degrade_reservation(self):
+
+        if len(self.moms) != 2:
+            cmt = "need 2 mom hosts: -p servers=<m1>,moms=<m1>:<m2>"
+            self.skip_test(reason=cmt)
```
Done

vchlum commented on test/tests/functional/pbs_maintenance_reservations.py:
```diff
@@ -36,60 +36,81 @@
+        """
+        resv_cmd = "%s -R %s -E %s %s --hosts %s" % \
+                   (self.rsub_cmd, start, end, others, " ".join(hosts))
+        cmd = "sudo -u %s %s" % (user, resv_cmd)
```
Fixed.

vchlum commented on test/tests/functional/pbs_maintenance_reservations.py:
```diff
@@ -36,60 +36,81 @@
+        self.rsub_cmd = os.path.join(self.server.pbs_conf[
+            'PBS_EXEC'], 'bin', 'pbs_rsub')
+
+    def now_plus_to_rsub_format(self, hours=0, minutes=0):
```
Fixed

prakashcv13 commented on src/cmds/pbs_rsub.c:
```diff
@@ -696,8 +765,158 @@ main(int argc, char *argv[], char *envp[])
+		pbs_statfree(bstat_head);	/* free info returned by pbs_stathost() */
+
+		pbs_errno = 0;
+		bstat_head = pbs_statvnode(connect, "", NULL, NULL);
```
instead of using pbs_statvnode(), we can use pbs_stathost(). We can make the code simpler instead of re-inventing the wheel. What do you think?

vchlum replied:
pbs_stathost() provides only list of hosts but we need the vnodes and its attributes because of crafting execvnodes. So, pbs_statvnode() is necessary.

I also use pbs_stathost() because we need to craft the select in the form like: "host=<host1>:ncpus=<ncpus_host1>...".

Maybe, we could use pbs_statvnode() and count the ncpus on all vnodes of the same host, but I think it would be more confusing than the used solution, and so it is better to call both pbs_stathost() and pbs_statvnode(). 

@prakashcv13 Do you think I should use only pbs_statvnode() and count the ncpus of the requested host from pbs_statvnode() (based on the same "host" attribute)?

prakashcv13 replied:
@vchlum - I should have spent more time in understanding why we are calling pbs_statvnode(). If we can avoid call to the pbs_stathost() and make the code better, we should do it. Could you please let me know why do you think using only pbs_statvnode() would be more confusing.

vchlum replied:
@prakashcv13 I thought this way there will be two similar blocks of code with easier reading, but I thought about that and you are right. I will remove the calling of pbs_stathost(). It is not a problem. We should avoid calling server if possible.

prakashcv13 commented on src/server/req_rescq.c:
```diff
@@ -222,6 +222,199 @@ cnvrt_timer_init()
+	char *end = NULL;
+	char *tmp_buf;
+	struct	resvinfo *rinfp, *prev;
+	attribute   tmpatr;
```
extra space after struct and attribute

vchlum replied:
Done.

prakashcv13 commented on src/server/req_rescq.c:
```diff
@@ -222,6 +222,199 @@ cnvrt_timer_init()
+ *
+ */
+void
+remove_node_from_resv(resc_resv *presv, struct pbsnode *pnode) {
```
'{' should be on the first column of the next line. Similar change is also needed in the newly added functions.

vchlum replied:
Done.

prakashcv13 commented on src/server/req_rescq.c:
```diff
@@ -222,6 +222,199 @@ cnvrt_timer_init()
+
+	strcpy(tmp_buf, pnode->nd_name);
+	tmp_buf[strlen(pnode->nd_name)] = ':';
+	tmp_buf[strlen(pnode->nd_name) + 1] = '\0';
```
we can use a single sprintf() call instead of the above 3 statements.

vchlum replied:
OK, done.

prakashcv13 commented on src/cmds/pbs_rsub.c:
```diff
@@ -696,8 +765,158 @@ main(int argc, char *argv[], char *envp[])
+							exit(2);
+						}
+						/* +14 stands for "+host=" + ":ncpus=" + '\0' */
+						select_str = (char*) realloc(select_str, strlen(tmp_str) + strlen(bstat->name) + strlen(ncpus) + 14);
```
I would suggest we initially malloc() a buffer of a fixed length say 500, and then use pbs_strcat() to concatenate subsequent values, instead of realloc()ing every time. This will save us from the calculations.

vchlum replied:
Definitely, good suggestion, done.

prakashcv13 commented on src/cmds/pbs_rsub.c:
```diff
@@ -696,8 +765,158 @@ main(int argc, char *argv[], char *envp[])
+				for (pattr = bstat->attribs; pattr; pattr = pattr->next) {
+				    if (pattr->resource && strcmp(pattr->name, "resources_available") == 0 && strcmp(pattr->resource, "ncpus") == 0)
+					ncpus = pattr->value;
+				}
```
wouldn't moving this for loop under "if (strcmp(bstat->name, *hostp) == 0)" would be better?

vchlum replied:
You were right, but this optimization is not feasible with the new implementation (pbs_stathost() is removed).

prakashcv13 commented on src/cmds/pbs_rsub.c:
```diff
@@ -696,8 +765,158 @@ main(int argc, char *argv[], char *envp[])
+			/* lets craft select_str */
+			for (bstat = bstat_head; bstat; bstat = bstat->next) {
+				for (pattr = bstat->attribs; pattr; pattr = pattr->next) {
+				    if (pattr->resource && strcmp(pattr->name, "resources_available") == 0 && strcmp(pattr->resource, "ncpus") == 0)
```
Let us use "ATTR_rescavail" macro instead of the string literal.

vchlum replied:
Done

prakashcv13 commented on src/cmds/pbs_rsub.c:
```diff
@@ -696,8 +765,158 @@ main(int argc, char *argv[], char *envp[])
+			exit(pbs_errno);
+		}
+
+		extend = strdup("m");
```
instead of calling strdup() and checking for it's success - let us use char *extend = "s".

vchlum replied:
OK, strcat() is used now.

vchlum commented on src/cmds/pbs_rsub.c:
```diff
@@ -696,8 +765,158 @@ main(int argc, char *argv[], char *envp[])
+
+		/* here, the select is crafted from hosts */
+		hostp = maintenance_hosts;
+		for (;*hostp ;hostp++) {
```
Fixed on all locations.

vchlum commented on src/cmds/pbs_rsub.c:
```diff
@@ -268,6 +277,32 @@ process_opts(int argc, char **argv, struct attrl **attrp, char *dest)
 
+	if (is_maintenance_resv) {
+		if (argc - optind > 0) {
+			maintenance_hosts = (char **)malloc(sizeof(char *) * (argc - optind + 1));
```
OK, done.

vchlum commented:
@prakashcv13 I addressed your comments. New ptl logs follow:
[ptl_reservations_output3.txt](https://github.com/PBSPro/pbspro/files/3137311/ptl_reservations_output3.txt)
[ptl_maintenance_output3.txt](https://github.com/PBSPro/pbspro/files/3137312/ptl_maintenance_output3.txt)

prakashcv13 commented on test/tests/functional/pbs_maintenance_reservations.py:
```diff
@@ -0,0 +1,513 @@
+                    'Resource_List.place': 'exclhost'}
+        self.server.expect(RESV, exp_attr, id=rid)
+
+    def test_maintenance_reconfirm_resvation_and_run(self):
```
resvation ==> reservation

vchlum replied:
The is typo fixed.

prakashcv13 commented on src/cmds/pbs_rsub.c:
```diff
@@ -696,8 +767,131 @@ main(int argc, char *argv[], char *envp[])
+					}
+					select_str_size = BUF_SIZE;
+
+					sprintf(select_str, "select=host=%s:ncpus=%d", *hostp, host_ncpus);
```
shall we use snprintf() instead of sprintf(), here and in other places?

vchlum replied:
snprintf() is used in all places now.

prakashcv13 commented:
Hi @vchlum - Thank you so much for making the changes. I have a few more comments.

vchlum commented on src/cmds/pbs_rsub.c:
```diff
@@ -696,8 +767,131 @@ main(int argc, char *argv[], char *envp[])
 
+	if (is_maintenance_resv) {
+		int i;
+		char tmp_str[BUF_SIZE];
```
I think 512 is acceptable. The worst case is 255 (for PBS_MAXHOSTNAME) + 10 (for max integer) + 13 (for "+host=:ncpus="), which is 278. Let me know if you still consider it to be wasting.

prakashcv13 replied:
hmmm. thank you for the explanation. let us keep it 512 bytes.

mike0042 replied:
The pbs_asprintf() function was introduced when we added "-Wall -Werror" to the compilation parameters. That might be another option in this case. Several buffer overrun issues were addressed with this.

vchlum replied:
pbs_asprintf() always allocates the memory.  Here, we repeatedly snprintf to the tmp_str in the for loop. Using pbs_asprintf() instead of snprintf, we would need to free the tmp_str after each use and it would be allocated with pbs_asprintf() again. Now, calculations are spared opposed to using pbs_asprintf(). I think that the string buffer that is large enough is more suitable in this case.

mike0042 replied:
Sounds reasonable. Thank you.

vchlum commented on src/cmds/pbs_rsub.c:
```diff
@@ -696,8 +767,131 @@ main(int argc, char *argv[], char *envp[])
+
+		/* here, the execvnodes is crafted from vnodes */
+		hostp = maintenance_hosts;
+		for (; *hostp; hostp++) {
```
@prakashcv13 If I understand correctly, you have suggested combining the crafting of 'select' and 'execvnodes' into one loop. It is possible but IMHO it will harm the readability pretty badly, and it will not improve the complexity. I would prefer keeping the crafting of 'select' and 'execvnodes' separately.

prakashcv13 replied:
@vchlum - I would vouch for performance instead of readability. What do you think?

vchlum replied:
This change is not included in the last commit yet. Let me know if you really prefer to combine the two for loops into one.

vchlum replied:
@prakashcv13 Well, I think the performance gain will be quite small, but no problem. OK, I will change it into one for loop in next commit. I will add some comments if necessary.

vchlum replied:
Done

vchlum commented on src/cmds/pbs_rsub.c:
```diff
@@ -788,98 +791,74 @@ main(int argc, char *argv[], char *envp[])
-						free(tmp_str);
-					}
+				if (strcmp(host, *hostp) == 0 && atoi(ncpus) > 0) {
+					host_ncpus += atoi(ncpus);
```
strtol() is used now.

vchlum commented on test/tests/functional/pbs_maintenance_reservations.py:
```diff
@@ -0,0 +1,513 @@
+        cmd = "%s -R %s -E %s %s --hosts %s" % \
+              (self.rsub_cmd, start, end, others, " ".join(hosts))
+
+        return self.du.run_cmd(self.server.hostname, cmd, runas=user)
```
PTL framework updated. I have added 'hosts' to the class Reservation().

vchlum commented on test/tests/functional/pbs_maintenance_reservations.py:
```diff
@@ -0,0 +1,513 @@
+        TestFunctional.setUp(self)
+
+        self.rsub_cmd = os.path.join(self.server.pbs_conf[
+            'PBS_EXEC'], 'bin', 'pbs_rsub')
```
OK, changed along with changing PTL framework.

vchlum commented on src/server/req_rescq.c:
```diff
@@ -222,6 +222,198 @@ cnvrt_timer_init()
+		return;
+	}
+
+	sprintf(tmp_buf, "%s:", pnode->nd_name);
```
It is used now.

vchlum commented on src/server/req_rescq.c:
```diff
@@ -222,6 +222,198 @@ cnvrt_timer_init()
+	struct resvinfo *rinfp, *prev;
+	attribute tmpatr;
+
+	tmp_buf = (char *)malloc(strlen(pnode->nd_name) + 2);
```
Done

vchlum commented on src/server/req_rescq.c:
```diff
@@ -222,6 +222,198 @@ cnvrt_timer_init()
+
+	tmp_buf = (char *)malloc(strlen(pnode->nd_name) + 2);
+	if (tmp_buf == NULL) {
+		snprintf(log_buffer, LOG_BUF_SIZE-1, "malloc failure (errno %d)", errno);
```
OK

vchlum commented on src/server/req_quejob.c:
```diff
@@ -2356,6 +2365,12 @@ req_resvSub(struct batch_request *preq)
+
+	if (is_maintenance) {
+		rid[0] = PBS_MNTNC_RESV_ID_CHAR;
+		qbuf[0] = PBS_MNTNC_RESV_ID_CHAR;
```
Done

vchlum commented on src/cmds/pbs_rsub.c:
```diff
@@ -696,8 +767,131 @@ main(int argc, char *argv[], char *envp[])
+						ncpus = pattr->value;
+				}
+
+				if (strcmp(host, *hostp) == 0 && atoi(ncpus) > 0){
```
Done

vchlum commented on src/cmds/pbs_rsub.c:
```diff
@@ -696,8 +767,131 @@ main(int argc, char *argv[], char *envp[])
+				}
+
+				if (strcmp(host, *hostp) == 0 && atoi(ncpus) > 0) {
+					host_ncpus += atoi(ncpus);
```
Removed.

vchlum commented:
Latest PTL log: 
[ptl_maintenance_output4.txt](https://github.com/PBSPro/pbspro/files/3142546/ptl_maintenance_output4.txt)

bhroam commented on src/cmds/pbs_rsub.c:
```diff
@@ -839,24 +848,33 @@ main(int argc, char *argv[], char *envp[])
+						snprintf(execvnodes_str, BUF_SIZE - 1, "(%s:ncpus=%d)", bstat->name, ncpus);
 					} else {
-						sprintf(tmp_str, "+(%s:ncpus=%s)", bstat->name, ncpus);
+						snprintf(tmp_str, BUF_SIZE - 1, "+(%s:ncpus=%d)", bstat->name, ncpus);
```
You don't need the -1.  snprintf() will only print N-1 bytes to leave space for the null byte.

vchlum replied:
Right, fixed.

bhroam commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -13872,10 +13874,11 @@ class Reservation(ResourceResv):
     dflt_attributes = {}
 
-    def __init__(self, username=None, attrs={}):
+    def __init__(self, username=None, attrs={}, hosts=[]):
```
It's dangerous to use mutable types as default arguments.  It means every call to this function will use the same memory.  If the list is modified in one call, it will still be modified in subsequent call.  It's safer to set it to None and then say if hosts is None: hosts = [].  I see that attrs in the same way, but it's also wrong.  Fix that one as well?

vchlum replied:
Both the attrs and hosts fixed.

bhroam commented on test/tests/functional/pbs_maintenance_reservations.py:
```diff
@@ -44,74 +44,62 @@ class TestMaintenanceReservations(TestFunctional):
+        try:
+            self.server.submit(r)
+        except PbsSubmitError as err:
+            msg = err.msg[0].strip()
```
Since maintenance reservations are handled in the pbs_rsub command, could you force PTL into CLI mode in setUp()?  This way if API mode is ever fixed, then this test won't fail.

vchlum replied:
Done

bhroam commented on src/cmds/pbs_rsub.c:
```diff
@@ -811,10 +810,35 @@ main(int argc, char *argv[], char *envp[])
+				/* here, the execvnodes is crafted */
+				if (strcmp(host, *hostp) == 0 && ncpus > 0) {
+					/* count ncpus of a host across vnodes
+					 * it will be used for crafting select */
```
Please move the */ to the line below

vchlum replied:
Done

prakashcv13 commented:
I have restarted debian and SUSE builds in travis as these builds had errored

prakashcv13 commented:
@vchlum - there is one comment from @mike0042 - that would need a reply or code changes. Could you please look into it. Thank you for your patience.

vchlum commented:
@prakashcv13 thx for the reminder. I replied to @mike0042. The debian build has succeeded but the suse build timeouts on retreiving the repository:-( Should I rebase?

prakashcv13 commented:
@vchlum - travis seems to be having an issue. Let us squash and rebase.

mike0042 commented on src/cmds/pbs_rsub.c:
```diff
@@ -100,7 +102,7 @@ process_opts(int argc, char **argv, struct attrl **attrp, char *dest)
 #endif
 
-	while ((c = getopt(argc, argv, "D:E:I:l:m:M:N:q:r:R:u:U:g:G:H:W:")) != EOF) {
+	while ((c = getopt(argc, argv, "D:E:I:l:m:M:N:q:r:R:u:U:g:G:H:W:-:")) != EOF) {
```
Shouldn't you be using getopt_long() for this? If you use getopt() then the following are equivalent:
pbs_rsub --hosts
pbs_rsub -- hosts

vchlum replied:
'pbs_rsub -- hosts' returns directive error. OK, I will change it to getopt_long() anyway.

mike0042 replied:
This will be the first time we've added a "long" parameter in PBS Pro. I suspect we will start using them more often, so thank you for making the change.

bhroam replied:
You can't use getopt_long().  That won't work on Windows.

vchlum replied:
Do we need to implement getopt_long() for Windows or will we use the current solution?

prakashcv13 replied:
As per comments from @mike0042, we need not implement getopt_long(). But, have we tested these changes on Windows?

vchlum replied:
No, I do not have any Windows machine. I probably could prepare some Windows testbed if necessary but it will take some time. I would appreciate help with the test on Windows.

mike0042 replied:
I don't want to force all this extra work on you @vchlum. I suggest we file a ticket to implement getopt_long() for Windows and refactor the existing getopt() calls in the future.

mike0042 commented on src/cmds/pbs_rsub.c:
```diff
@@ -256,6 +258,13 @@ process_opts(int argc, char **argv, struct attrl **attrp, char *dest)
 
+			case '-':
+				if (strcmp(optarg, "hosts") == 0)
+					is_maintenance_resv = 1;
```
Will this impact all vnodes assigned to the list of hosts?

vchlum replied:
The vnodes are obtained later with pbs_statvnode(). For each host passed as a parameter (here), the associated vnodes are found and used in crafting execvnodes.

prakashcv13 commented:
I restarted the SUSE build and travis is green now.

bhroam commented:
Thanks for the nice new feature, @vchlum.  I merged in the PR.
